### PR TITLE
Add cinn_instruction_run_op for launching execution of a cinn instruction

### PR DESCRIPTION
**Pull request by CtfGo**
### PR types

Performance optimization
### PR changes

OPs
### Describe

1. add a new operator `cinn_instruction_run_op`: detail explanation of this operator can refer to the comments in its OpMaker.
2. upgrade the organization structure of the cache in `CinnCompiler`: add a `index->cached object` map as the real storage of all `CinnCompiledObjects` and the old two cache maps save their indexes as value, and add a new interface `GetCompiledObject` for getting the cached result by a index.
3. add a new reverse name map `cinn variable name -> paddle variable name` in `CinnLaunchContext`, which is generated by the current map `paddle variable name -> cinn variable name` that is also extended to include the internal variables(the map only include external variables originally).
4. add a new interface `GetCinnBufferOfVar` in `CinnLaunchContext` for getting the `cinn_buffer_t` object of a specified variable.

### CODE REVIEW

File: paddle/fluid/operators/cinn/CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
-register_operators(EXCLUDES cinn_launch_op)
 
+cc_library(cinn_op_helper SRCS cinn_op_helper.cc DEPS operator device_context)
 cc_library(cinn_launch_context SRCS cinn_launch_context.cc DEPS ddim lod_tensor scope cinn)
-op_library(cinn_launch_op SRCS cinn_launch_op.cc cinn_launch_op.cu.cc DEPS string_helper cinn cinn_compiler cinn_launch_context)
+
+SET(CINN_OP_DEPS string_helper cinn cinn_compiler cinn_op_helper cinn_launch_context)
+register_operators(DEPS ${CINN_OP_DEPS})
+#op_library(cinn_launch_op SRCS cinn_launch_op.cc cinn_launch_op.cu.cc DEPS string_helper cinn cinn_compiler cinn_op_helper cinn_launch_context)
+#op_library(cinn_instruction_run_op SRCS cinn_instruction_run_op.cc cinn_instruction_run_op.cu.cc DEPS string_helper cinn cinn_compiler cinn_op_helper cinn_launch_context)
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/CMakeLists.txt]: 如果这两句无用请直接删除，请不要在此注释。

File: paddle/fluid/operators/cinn/cinn_launch_context.cc [hunk 2]
```diff
@@ -189,6 +208,20 @@ CinnLaunchContext::FinalizeArguments() const {
...
 }
 
+cinn_buffer_t* CinnLaunchContext::GetCinnBufferOfVar(
+    const std::string& paddle_var_name) {
+  auto res = paddle2cinn_varmap_.find(paddle_var_name);
+  PADDLE_ENFORCE_NE(
+      res, paddle2cinn_varmap_.end(),
+      platform::errors::InvalidArgument(
+          "Variable(%s) not found in compilation result", paddle_var_name));
+  auto it = name2argument_.find(res->second);
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.cc:218]: 请在注释中注明`name2argument_`中的key使用的是`cinn var`的名字。

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                          "Index(%ld) > instructions.size(%ld).", ins_index,
+                          instructions.size()));
+    auto&& instruction = instructions.at(ins_index);
+
+    // step 2: prepare the input and output arguments of the instruction
+    details::CinnLaunchContext* launch_context =
+        compiled_object.launch_context.get();
+    auto share_argument_buffer_fn = [launch_context,
+                                     &ctx](const std::string& var_name) {
+      cinn_buffer_t* buffer = launch_context->GetCinnBufferOfVar(var_name);
```
Comment (thisjiang) [inline paddle/fluid/operators/cinn/cinn_instruction_run_op.h:57]: 这儿是否可能返回为空指针或者野指针啊？特别是在多线程环境下？建议加个`PADDLE_ENFORCE_NE`

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      framework::Variable* var = ctx.scope().GetVar(var_name);
```
Comment (thisjiang) [inline paddle/fluid/operators/cinn/cinn_instruction_run_op.h:58]: 同理，这儿要不先判断下`ctx.scope().HasVar(var_name)`（如果有这个接口）？

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      auto* tensor = var->template GetMutable<framework::LoDTensor>();
+      buffer->memory =
+          reinterpret_cast<uint8_t*>(tensor->mutable_data<T>(ctx.GetPlace()));
```
Comment (thisjiang) [inline paddle/fluid/operators/cinn/cinn_instruction_run_op.h:61]: 这儿为啥要转为`uint8_t*`啊？CINN的buffer指针都是`uint8_t*`类型的么？

File: paddle/fluid/operators/cinn/CMakeLists.txt (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/CMakeLists.txt]: done

File: paddle/fluid/operators/cinn/cinn_launch_context.cc (hunk 2 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_launch_context.cc:218]: 已在头文件声明处添加注释

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h (hunk 3 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_instruction_run_op.h:57]: GetCinnBufferOfVar函数中有 ENFORCE 判断，若是变量名不存在会报错

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h (hunk 4 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_instruction_run_op.h:58]: scope.GetVar 函数中有 ENFORCE 判断

File: paddle/fluid/operators/cinn/cinn_instruction_run_op.h (hunk 5 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_instruction_run_op.h:61]: 裸指针，是。cinn 会根据实际数据类型读取

File: paddle/fluid/operators/cinn/CMakeLists.txt (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/CMakeLists.txt]: done

**Review by thisjiang — APPROVED** [review]
LGTM

**Review by wzzju — APPROVED** [review]
LGTM

**Pull request merged**